suave/core/engine.go

```diff
@@ -284,71 +254,106 @@ func (e *ConfidentialStoreEngine) NewMessage(message DAMessage) error {
-		if !errors.Is(err, ErrBidAlreadyPresent) {
-			return fmt.Errorf("unexpected error while initializing bid from transport: %w", err)
+		bidBytes, err := SerializeBidForSigning(&sw.Bid)
+		if err != nil {
+			return fmt.Errorf("confidential engine: could not hash received bid: %w", err)
+		}
+		recoveredBidSigner, err := e.daSigner.Sender(bidBytes, sw.Bid.Signature)
+		if err != nil {
+			return fmt.Errorf("confidential engine: incorrect bid signature: %w", err)
+		}
+		expectedBidSigner, err := ExecutionNodeFromTransaction(sw.Bid.CreationTx)
+		if err != nil {
+			return fmt.Errorf("confidential engine: could not recover signer from bid: %w", err)
+		}
+		if recoveredBidSigner != expectedBidSigner {
+			return fmt.Errorf("confidential engine: bid signer %x, expected %x", recoveredBidSigner, expectedBidSigner)
+		}
+
+		if !slices.Contains(sw.Bid.AllowedStores, recoveredMessageSigner) {
+			return fmt.Errorf("confidential engine: sw signer %x not allowed to store on bid %x", recoveredMessageSigner, sw.Bid.Id)
+		}
+
+		if !slices.Contains(sw.Bid.AllowedPeekers, sw.Caller) {
+			return fmt.Errorf("confidential engine: caller %x not allowed on bid %x", sw.Caller, sw.Bid.Id)
+		}
+
+		// TODO: move to types.Sender()
+		_, err = e.chainSigner.Sender(sw.Bid.CreationTx)
+		if err != nil {
+			return fmt.Errorf("confidential engine: creation tx for bid id %x is not signed properly: %w", sw.Bid.Id, err)
 		}
 	}
 
-	_, err = e.backend.Store(message.Bid, message.Caller, message.Key, message.Value)
-	if err != nil {
-		return fmt.Errorf("unexpected error while storing: %w", err)
+	for _, sw := range message.StoreWrites {
+		err = e.backend.InitializeBid(sw.Bid)
+		if err != nil {
+			if !errors.Is(err, ErrBidAlreadyPresent) {
+				log.Error("confidential engine: unexpected error while initializing bid from transport: %w", err)
+				continue // Don't abandon!
+			}
+		}
+
+		_, err = e.backend.Store(sw.Bid, sw.Caller, sw.Key, sw.Value)
+		if err != nil {
+			log.Error("confidential engine: unexpected error while storing: %w", err)
+			continue // Don't abandon!
+		}
 	}
 
 	return nil
 }
 
-func SerializeBidForSigning(bid Bid) ([]byte, error) {
-	bid.Signature = []byte{}
-
-	bidBytes, err := json.Marshal(bid)
+func SerializeBidForSigning(bid *Bid) ([]byte, error) {
+	bidBytes, err := json.Marshal(Bid{
+		Id:                  bid.Id,
+		Salt:                bid.Salt,
+		DecryptionCondition: bid.DecryptionCondition,
+		AllowedPeekers:      bid.AllowedPeekers,
+		AllowedStores:       bid.AllowedStores,
+		Version:             bid.Version,
+		CreationTx:          bid.CreationTx,
+	})
 	if err != nil {
 		return []byte{}, err
 	}
 
 	return []byte(fmt.Sprintf("\x19Suave Signed Message:\n%d%s", len(bidBytes), string(bidBytes))), nil
 }
 
-func SerializeMessageForSigning(message DAMessage) ([]byte, error) {
-	message.Signature = []byte{}
-
-	msgBytes, err := json.Marshal(message)
+func SerializeMessageForSigning(message *DAMessage) ([]byte, error) {
```

Comment:
```suggestion
func (message *DAMessage) SerializeForSigning() ([]byte, error) {
```
We could do a common interface for objects that get signed.

Reply:
We could, on the other hand I don't think this is too much